### DISCUSSION TITLE
VSCode extension ideas

### DISCUSSION THREAD

**Discussion by bugarela** [informalsystems/quint#254] [Ideas]
I'm creating this discussion to keep track of all our ideas for the VSCode extension. Let's try to keep one idea per top-level comment to keep things organized.

**Comment by bugarela** [comment]
**Highlight actions**

We could use [semantic highlighting](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide) to have different fontfaces for actions by consulting the effects. I'm not sure how to make that not look completely messy, we should be careful, but I thin that might be very useful if we get it right.

**Comment by bugarela** [comment]
**Tasks**

It is possible to read arbitrary files and detect tasks, and these tasks can be run from the VSCode command pallete. Imagine we have some mdx-like (or some other configuration format like .toml files as atomkraft is using), detect tasks from that and have commands like `Check liveness property`, `Simulate malicious example`, etc. Those can be completely spec-specific since they are defined by a configuration file, as the example is specific to Quint typescript project.
 
![image](https://user-images.githubusercontent.com/18356186/193144368-e21059a5-9410-4623-a95e-dd0c5e7d7d59.png)

**Comment by bugarela** [comment]
**Notebooks**

For protocol specification, we can have a markdown renderer that parses and processes Quint code blocks. The original `.md` file should be readable by itself so people don't depend on the plugin to read the file.

![image](https://user-images.githubusercontent.com/18356186/193144451-63bd98c6-02ff-4546-aa6c-ed356b334c43.png)

**Reply by konnov** [reply]
How about the opposite: The users could write markdown in Quint comments. We could also render that in nice HTML. This would look like literate programming.

**Comment by bugarela** [comment]
**Code completion**

Using type information + dot call format, we can offer [code completion](https://code.visualstudio.com/api/language-extensions/programmatic-language-features#show-code-completion-proposals) like any OOP language. That is, given a set `S`, typing `S.` shows operators that take sets as their first argument.

Example from vscode docs:
![image](https://code.visualstudio.com/assets/api/language-extensions/language-support/code-completion.gif)

**Comment by bugarela** [comment]
**Type hints**

After opening parenthesis, we can show an operator's type (and maybe effect?) signature to help users remember which arguments are needed. PS: Currently, our built-in operators don't have names for their arguments, but it's much helpful to show signatures with named arguments (`index: int` as opposed to only `int`)

Image from [vscode docs](https://code.visualstudio.com/api/language-extensions/programmatic-language-features#help-with-function-and-method-signatures)
![image](https://code.visualstudio.com/assets/api/language-extensions/language-support/signature-help.gif)

**Reply by bugarela** [reply]
Instead of showing effects on hover, we can actually show the mode, as typescript shows `const` before the name in this example:
![image](https://user-images.githubusercontent.com/18356186/193265861-80ba53cf-c6ef-44d3-bc2b-d5c45f303c68.png)

Actions can show the variables they update with the `action<...>` syntax

**Comment by bugarela** [comment]
**Type annotation inference** 

We can infer types for the operators, but as in many programming languages, it's nice if the user explicitly annotates top-level operators so they can be sure their intentions are met. We can automate that task by providing a button for annotating an operator with it's inferred type, so the user only has to check for it's correction and then benefit from a established type contract.

![Peek 2022-09-29 18-44](https://user-images.githubusercontent.com/18356186/193147714-a78fa9e0-e6be-488d-beea-ca99ac1ef429.gif)

PS: idk why my gif recorder is so dark, sorry

**Reply by konnov** [reply]
This would be a really cool feature!

**Comment by bugarela** [comment]
**Code Navigation**

Easily find where a symbol is defined or all references to a symbol.

Example from [vscode docs](https://code.visualstudio.com/api/language-extensions/programmatic-language-features#show-definitions-of-a-symbol):
![image](https://code.visualstudio.com/assets/api/language-extensions/language-support/goto-definition.gif)

**Reply by bugarela** [reply]
Done in https://github.com/informalsystems/quint/pull/546

**Comment by bugarela** [comment]
Maybe we can render diagrams for individual actions, similar to what the ANTLR extension does

![image](https://user-images.githubusercontent.com/18356186/207311316-863b6f28-c52a-4313-a736-675991dd8380.png)

**Reply by konnov** [reply]
This should be relatively easy to implement now, since we have the call graph. Though it may be that people would find "Find all references" more useful than the graph (they are hard to render and hard to navigate).

**Reply by bugarela** [reply]
I was not thinking about this in terms of the call graph, but in terms of state machine transitions. I'm not sure if it is feasible, just raising the idea. For the call graph, I agree that "Go to References" is more useful.

**Comment by shonfeder** [comment]
Under the heading of support for literate programming.

@konnov noted that there are some standard practices spec authors (in particular
protocol designers) can adopt that would improve the quality usefulness of
specs. E.g.,

- Examples (using `run`, so these also amount to tests of the spec)
- Invariants
- Formalization in quint (as opposed to only describing something in a paragraph)

Support for literate formal specs could include templates and linting to
automate encouragement of enforcement of these practices.
